### PR TITLE
Add owner_address_hash and other fields to token_instances; Move toke…

### LINKED ISSUE #8231: Token ids for ERC-721
https://github.com/blockscout/blockscout/issues/3050
https://github.com/blockscout/blockscout/issues/6068
https://github.com/blockscout/blockscout/issues/1074

### PR DESCRIPTION
**Pull request by nikitosing**
…n instances' insert into synchronous block import stage
Close #8231 

## Changelog
- Add  `owner_address_hash`, `owner_updated_at_block`, `owner_updated_at_log_index` to the `token_instances` table
- These fields updates on token transfers import
- To backfill indexed token transfers and instances, was created `Explorer.TokenInstanceOwnerAddressMigration` process.
- Implementation of `Indexer.Fetcher.TokenInstance.Sanitize` was changed due to migration to the new approach of token instance indexing. Old implementation moved to `Indexer.Fetcher.TokenInstance.LegacySanitize`. Envs updated accordingly
https://github.com/blockscout/docs/pull/181
## Checklist for your Pull Request (PR)

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [ ] If I added new functionality, I added tests covering it.
  - [ ] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [ ] I checked whether I should update the docs and did so by submitting a PR to https://github.com/blockscout/docs
  - [x] If I added/changed/removed ENV var, I submitted a PR to https://github.com/blockscout/docs to update the list of env vars at https://github.com/blockscout/docs/blob/master/for-developers/information-and-settings/env-variables.md and I updated the version to `master` in the Version column. Changes will be reflected in this table: https://docs.blockscout.com/for-developers/information-and-settings/env-variables. 
  - [ ] If I add new indices into DB, I checked, that they are not redundant with PGHero or other tools

### CODE REVIEW

**Review by Qwerty5Uiop — APPROVED** [review]

File: apps/explorer/lib/explorer/token_instance_owner_address_migration/helper.ex [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+  import Ecto.Query,
+    only: [
+      from: 2
+    ]
+
+  alias Explorer.{Chain, Repo}
+  alias Explorer.Chain.Token.Instance
+  alias Explorer.Chain.TokenTransfer
+
+  {:ok, burn_address_hash} = Chain.string_to_address_hash("0x0000000000000000000000000000000000000000")
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/token_instance_owner_address_migration/helper.ex:14]: We have `import Explorer.Chain.SmartContractburn_address_hash_string()` for burn address definition

File: apps/explorer/lib/explorer/chain.ex [hunk 2]
```diff
@@ -4391,12 +4391,12 @@ defmodule Explorer.Chain do
...
   end
 
-  @spec stream_unfetched_token_instances(
+  @spec stream_not_inserted_token_instances(
           initial :: accumulator,
           reducer :: (entry :: map(), accumulator -> accumulator)
         ) :: {:ok, accumulator}
         when accumulator: term()
-  def stream_unfetched_token_instances(initial, reducer) when is_function(reducer, 2) do
+  def stream_not_inserted_token_instances(initial, reducer) when is_function(reducer, 2) do
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain.ex:4399]: Please add doc description for the function

File: apps/explorer/lib/explorer/chain.ex [hunk 3]
```diff
@@ -4438,6 +4438,21 @@ defmodule Explorer.Chain do
     Repo.stream_reduce(distinct_query, initial, reducer)
   end
 
+  @spec stream_token_instances_with_unfetched_metadata(
+          initial :: accumulator,
+          reducer :: (entry :: map(), accumulator -> accumulator)
+        ) :: {:ok, accumulator}
+        when accumulator: term()
+  def stream_token_instances_with_unfetched_metadata(initial, reducer) when is_function(reducer, 2) do
```
Comment (vbaranov) [inline apps/explorer/lib/explorer/chain.ex:4446]: Please add doc description for the function

File: apps/explorer/lib/explorer/chain.ex [hunk 4]
```diff
@@ -4696,18 +4717,37 @@ defmodule Explorer.Chain do
...
+          metadata: fragment("EXCLUDED.metadata"),
+          error: fragment("EXCLUDED.error"),
+          owner_updated_at_block: fragment("?", token_instance.owner_updated_at_block),
+          owner_updated_at_log_index: fragment("?", token_instance.owner_updated_at_log_index),
+          owner_address_hash: fragment("?", token_instance.owner_address_hash),
+          inserted_at: fragment("LEAST(?, EXCLUDED.inserted_at)", token_instance.inserted_at),
+          updated_at: fragment("GREATEST(?, EXCLUDED.updated_at)", token_instance.updated_at)
+        ]
+      ],
+      where: fragment("? IS NULL", token_instance.metadata)
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain.ex:4747]: Do we really need to use `fragment` here?

File: apps/explorer/lib/explorer/chain/import/runner/blocks.ex [hunk 5]
```diff
@@ -577,6 +596,155 @@ defmodule Explorer.Chain.Import.Runner.Blocks do
...
+    )
+  end
+
+  defp token_instances_on_conflict do
+    from(
+      token_instance in Instance,
+      update: [
+        set: [
+          metadata: fragment("?", token_instance.metadata),
+          error: fragment("?", token_instance.error),
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain/import/runner/blocks.ex:737]: Same here

File: apps/explorer/lib/explorer/chain/import/runner/token_instances.ex [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+      )
+  end
+
+  defp default_on_conflict do
+    from(
+      token_instance in TokenInstance,
+      update: [
+        set: [
+          metadata: fragment("?", token_instance.metadata),
+          error: fragment("?", token_instance.error),
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain/import/runner/token_instances.ex:90]: Same here

File: apps/explorer/lib/explorer/chain/import/runner/token_instances.ex [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        ]
+      ],
+      where:
+        fragment("EXCLUDED.owner_address_hash IS NOT NULL") and fragment("EXCLUDED.owner_updated_at_block IS NOT NULL") and
+          (fragment("EXCLUDED.owner_updated_at_block > ?", token_instance.owner_updated_at_block) or
+             (fragment("EXCLUDED.owner_updated_at_block = ?", token_instance.owner_updated_at_block) and
+                fragment("EXCLUDED.owner_updated_at_log_index >= ?", token_instance.owner_updated_at_log_index)) or
+             fragment("? IS NULL", token_instance.owner_updated_at_block) or
+             fragment("? IS NULL", token_instance.owner_address_hash))
+    )
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain/import/runner/token_instances.ex:105]: Same here

File: apps/explorer/lib/explorer/chain/token_transfer.ex [hunk 8]
```diff
@@ -360,4 +360,12 @@ defmodule Explorer.Chain.TokenTransfer do
   end
 
   def filter_by_type(query, _), do: query
+
+  def only_consensus_transfers_query do
+    from(token_transfer in __MODULE__,
+      inner_join: block in Block,
+      on: token_transfer.block_hash == block.hash,
+      where: block.consensus
```
Comment (sl1depengwyn) [inline apps/explorer/lib/explorer/chain/token_transfer.ex:368]: According to https://github.com/blockscout/blockscout/commit/2981a6455ec7f064546fb657982b83a039cb446a I think it's better to do `block.consensus == true`

**Review by sl1depengwyn — APPROVED** [review]

**Review by vbaranov — APPROVED** [review]

**Pull request merged**